// vant/lib/sku/components/SkuRowItem.js
'use strict';

exports.__esModule = true;

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _filter = require('lodash/filter');

var _filter2 = _interopRequireDefault(_filter);

var _keys = require('lodash/keys');

var _keys2 = _interopRequireDefault(_keys);

var _every = require('lodash/every');

var _every2 = _interopRequireDefault(_every);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  render: function render() {
    var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _vm.isChoosable ? _c('span', { class: { 'van-sku-row__item': true, 'van-sku-row__item--active': _vm.isChoosed }, on: { "click": _vm.onSkuSelected } }, [_vm._v("\n  " + _vm._s(_vm.skuValue.name) + "\n")]) : _c('span', { staticClass: "van-sku-row__item van-sku-row__item--disabled" }, [_vm._v(_vm._s(_vm.skuValue.name))]);
  },

  name: 'van-sku-row-item',

  props: {
    skuEventBus: Object,
    skuValue: Object,
    skuList: Array,
    selectedSku: Object,
    skuKeyStr: String
  },

  computed: {
    isChoosed: function isChoosed() {
      return this.skuValue.id === this.selectedSku[this.skuKeyStr];
    },
    isChoosable: function isChoosable() {
      var _assign;

      var matchedSku = (0, _assign3.default)({}, this.selectedSku, (_assign = {}, _assign[this.skuKeyStr] = this.skuValue.id, _assign));
      var skusToCheck = (0, _filter2.default)((0, _keys2.default)(matchedSku), function (skuKey) {
        return matchedSku[skuKey] !== '';
      });
      var filteredSku = (0, _filter2.default)(this.skuList, function (sku) {
        return (0, _every2.default)(skusToCheck, function (skuKey) {
          // 后端给的skuValue.id有时候是数字有时候是字符串，全等会匹配不上
          return matchedSku[skuKey] == sku[skuKey]; // eslint-disable-line
        });
      });
      var stock = filteredSku.reduce(function (total, sku) {
        return total += sku.stock_num;
      }, 0);

      return stock > 0;
    }
  },

  methods: {
    onSkuSelected: function onSkuSelected() {
      this.skuEventBus.$emit('sku:select', (0, _assign3.default)({}, this.skuValue, { skuKeyStr: this.skuKeyStr }));
    }
  }
};
