// vant/lib/sku/components/SkuRow.js
'use strict';

exports.__esModule = true;

var _components;

var _SkuRowItem = require('./SkuRowItem');

var _SkuRowItem2 = _interopRequireDefault(_SkuRowItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  render: function render() {
    var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { staticClass: "van-sku-row" }, [_c('div', { staticClass: "van-sku-row__title" }, [_vm._v(_vm._s(_vm.skuRow.k) + "：")]), _c('div', { staticClass: "van-sku-row__items" }, [_vm._t("sku-row-items", _vm._l(_vm.skuRow.v, function (skuValue, index) {
      return _c('van-sku-row-item', { key: index, attrs: { "skuKeyStr": _vm.skuRow.k_s, "skuValue": skuValue, "skuEventBus": _vm.$attrs.skuEventBus, "selectedSku": _vm.$attrs.selectedSku, "skuList": _vm.$attrs.skuList } });
    }))], 2)]);
  },

  name: 'van-sku-row',

  inheritAttrs: false,

  components: (_components = {}, _components[_SkuRowItem2.default.name] = _SkuRowItem2.default, _components),

  props: {
    skuRow: Object
  }
};
