// vant/lib/sku/utils/skuHelper.js
'use strict';

exports.__esModule = true;
exports.getSelectedSkuValues = exports.getSkuComb = exports.isAllSelected = exports.normalizeSkuTree = undefined;

var _isArray = require('lodash/isArray');

var _isArray2 = _interopRequireDefault(_isArray);

var _keys = require('lodash/keys');

var _keys2 = _interopRequireDefault(_keys);

var _filter = require('lodash/filter');

var _filter2 = _interopRequireDefault(_filter);

var _find = require('lodash/find');

var _find2 = _interopRequireDefault(_find);

var _isObject = require('lodash/isObject');

var _isObject2 = _interopRequireDefault(_isObject);

var _every = require('lodash/every');

var _every2 = _interopRequireDefault(_every);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
  normalize sku tree

  [
    {
      count: 2,
      k: "品种", // 规格名称 skuKeyName
      k_id: "1200", // skuKeyId
      k_s: "s1" // skuKeyStr
      v: [ // skuValues
        { // skuValue
          id: "1201", // skuValueId
          name: "萌" // 具体的规格值 skuValueName
        }, {
          id: "973",
          name: "帅"
        }
      ]
    },
    ...
  ]
                |
                v
  {
    s1: [{
      id: "1201",
      name: "萌"
    }, {
      id: "973",
      name: "帅"
    }],
    ...
  }
 */
var normalizeSkuTree = exports.normalizeSkuTree = function normalizeSkuTree(skuTree) {
  var normalizedTree = {};
  skuTree.forEach(function (treeItem) {
    normalizedTree[treeItem.k_s] = treeItem.v;
  });
  return normalizedTree;
};

// 判断是否所有的sku都已经选中
var isAllSelected = exports.isAllSelected = function isAllSelected(skuTree, selectedSku) {
  if (!(0, _isArray2.default)(skuTree)) {
    throw new Error('skuTree must be array');
  }
  if (!(0, _isObject2.default)(selectedSku)) {
    throw new Error('selectedSku must be object');
  }

  var selected = (0, _filter2.default)((0, _keys2.default)(selectedSku), function (skuKeyStr) {
    return selectedSku[skuKeyStr] !== '';
  });
  return skuTree.length === selected.length;
};

// 根据已选择的sku获取skuComb
var getSkuComb = exports.getSkuComb = function getSkuComb(skuList, selectedSku) {
  var skuComb = (0, _find2.default)(skuList, function (skuComb) {
    return (0, _every2.default)((0, _keys2.default)(selectedSku), function (skuKeyStr) {
      // 后端给的key有时数字有时字符串，需要兼容=。=
      return skuComb[skuKeyStr] == selectedSku[skuKeyStr]; // eslint-disable-line
    });
  });
  return skuComb;
};

// 获取已选择的sku名称
var getSelectedSkuValues = exports.getSelectedSkuValues = function getSelectedSkuValues(skuTree, selectedSku) {
  var normalizedTree = normalizeSkuTree(skuTree);
  return (0, _keys2.default)(selectedSku).reduce(function (selectedValues, skuKeyStr) {
    var skuValues = normalizedTree[skuKeyStr];
    var skuValueId = selectedSku[skuKeyStr];

    if (skuValueId) {
      var skuValue = (0, _find2.default)(skuValues, function (skuValue) {
        return skuValue.id === skuValueId;
      });
      skuValue && selectedValues.push(skuValue);
    }
    return selectedValues;
  }, []);
};

var SkuHelper = {
  normalizeSkuTree: normalizeSkuTree,
  isAllSelected: isAllSelected,
  getSkuComb: getSkuComb,
  getSelectedSkuValues: getSelectedSkuValues
};
exports.default = SkuHelper;
