// zan-utils/lib/date/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './formatDate', './getCurrentDay', './getCurrentWeek', './getValidDate', './makeDateStr', './makeDateTimeStr', './parseDate', './travel', './compare'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./formatDate'), require('./getCurrentDay'), require('./getCurrentWeek'), require('./getValidDate'), require('./makeDateStr'), require('./makeDateTimeStr'), require('./parseDate'), require('./travel'), require('./compare'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.formatDate, global.getCurrentDay, global.getCurrentWeek, global.getValidDate, global.makeDateStr, global.makeDateTimeStr, global.parseDate, global.travel, global.compare);
    global.index = mod.exports;
  }
})(this, function (module, exports, _formatDate, _getCurrentDay, _getCurrentWeek, _getValidDate, _makeDateStr, _makeDateTimeStr, _parseDate, _travel, _compare) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _formatDate2 = _interopRequireDefault(_formatDate);

  var _getCurrentDay2 = _interopRequireDefault(_getCurrentDay);

  var _getCurrentWeek2 = _interopRequireDefault(_getCurrentWeek);

  var _getValidDate2 = _interopRequireDefault(_getValidDate);

  var _makeDateStr2 = _interopRequireDefault(_makeDateStr);

  var _makeDateTimeStr2 = _interopRequireDefault(_makeDateTimeStr);

  var _parseDate2 = _interopRequireDefault(_parseDate);

  var _travel2 = _interopRequireDefault(_travel);

  var _compare2 = _interopRequireDefault(_compare);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = {
    formatDate: _formatDate2['default'],
    getCurrentDay: _getCurrentDay2['default'],
    getCurrentWeek: _getCurrentWeek2['default'],
    getValidDate: _getValidDate2['default'],
    makeDateStr: _makeDateStr2['default'],
    makeDateTimeStr: _makeDateTimeStr2['default'],
    parseDate: _parseDate2['default'],
    travel: _travel2['default'],
    compare: _compare2['default']
  };
  module.exports = exports['default'];
});
