// vant/lib/utils/clickoutside.js
'use strict';

exports.__esModule = true;

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var isServer = _vue2.default.prototype.$isServer; /**
                                                   * v-clickoutside
                                                   * @desc 点击元素外面才会触发的事件
                                                   * @example
                                                   * ```vue
                                                   * <div v-clickoutside="handleClose">
                                                   * ```
                                                   */

var clickoutsideContext = '@@clickoutsideContext';

exports.default = {
  bind: function bind(el, binding, vnode) {
    var documentHandler = function documentHandler(e) {
      if (vnode.context && !el.contains(e.target)) {
        vnode.context[el[clickoutsideContext].methodName]();
      }
    };
    el[clickoutsideContext] = {
      documentHandler: documentHandler,
      methodName: binding.expression,
      arg: binding.arg || 'click'
    };
    !isServer && document.addEventListener(el[clickoutsideContext].arg, documentHandler);
  },
  update: function update(el, binding) {
    el[clickoutsideContext].methodName = binding.expression;
  },
  unbind: function unbind(el) {
    !isServer && document.removeEventListener(el[clickoutsideContext].arg, el[clickoutsideContext].documentHandler);
  },
  install: function install(Vue) {
    Vue.directive('clickoutside', {
      bind: this.bind,
      unbind: this.unbind
    });
  }
};
