// zan-utils/lib/money/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './format', './formatLargeNumber', './add', './minus', './multiply', './toCent'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./format'), require('./formatLargeNumber'), require('./add'), require('./minus'), require('./multiply'), require('./toCent'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.format, global.formatLargeNumber, global.add, global.minus, global.multiply, global.toCent);
    global.index = mod.exports;
  }
})(this, function (module, exports, _format, _formatLargeNumber, _add, _minus, _multiply, _toCent) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _format2 = _interopRequireDefault(_format);

  var _formatLargeNumber2 = _interopRequireDefault(_formatLargeNumber);

  var _add2 = _interopRequireDefault(_add);

  var _minus2 = _interopRequireDefault(_minus);

  var _multiply2 = _interopRequireDefault(_multiply);

  var _toCent2 = _interopRequireDefault(_toCent);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = {
    format: _format2['default'],
    formatLargeNumber: _formatLargeNumber2['default'],
    add: _add2['default'],
    minus: _minus2['default'],
    multiply: _multiply2['default'],
    toCent: _toCent2['default']
  };
  module.exports = exports['default'];
});
