// vant/lib/mixins/popup/popup-manager.js
'use strict';

exports.__esModule = true;

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

var _popupContext = require('./popup-context');

var _popupContext2 = _interopRequireDefault(_popupContext);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getModal = function getModal() {
  if (_vue2.default.prototype.$isServer) return;
  var modalDom = _popupContext2.default.getContext('modalDom');

  if (modalDom) {
    _popupContext2.default.setContext('hasModal', true);
  } else {
    _popupContext2.default.setContext('hasModal', false);

    modalDom = document.createElement('div');
    _popupContext2.default.setContext('modalDom', modalDom);

    modalDom.addEventListener('touchmove', function (event) {
      event.preventDefault();
      event.stopPropagation();
    });

    modalDom.addEventListener('click', function () {
      PopupManager.handleOverlayClick && PopupManager.handleOverlayClick();
    });
  }

  return modalDom;
};

var PopupManager = {
  nextZIndex: function nextZIndex() {
    return _popupContext2.default.plusKeyByOne('zIndex');
  },
  getInstance: function getInstance(id) {
    return _popupContext2.default.getContext('instances')[id];
  },
  register: function register(id, instance) {
    if (id && instance) {
      var instances = _popupContext2.default.getContext('instances');
      instances[id] = instance;
    }
  },
  deregister: function deregister(id) {
    if (id) {
      var instances = _popupContext2.default.getContext('instances');
      instances[id] = null;
      delete instances[id];
    }
  },


  /**
   * 遮罩层点击回调，`closeOnClickOverlay`为`true`时会关闭当前`popup`
   */
  handleOverlayClick: function handleOverlayClick() {
    var modalStack = _popupContext2.default.getContext('modalStack');
    var topModal = modalStack[modalStack.length - 1];
    if (!topModal) return;

    var instance = PopupManager.getInstance(topModal.id);
    if (instance && instance.closeOnClickOverlay) {
      instance.close();
    }
  },
  openModal: function openModal(id, zIndex, dom) {
    if (!id || zIndex === undefined) return;

    var modalStack = _popupContext2.default.getContext('modalStack');

    for (var i = 0, len = modalStack.length; i < len; i++) {
      var item = modalStack[i];
      if (item.id === id) {
        return;
      }
    }

    var modalDom = getModal();

    modalDom.classList.add('van-modal');

    var domParentNode = void 0;
    if (dom && dom.parentNode && dom.parentNode.nodeType !== 11) {
      domParentNode = dom.parentNode;
    } else {
      domParentNode = document.body;
    }
    domParentNode.appendChild(modalDom);

    if (zIndex) {
      modalDom.style.zIndex = zIndex;
    }
    modalDom.style.display = '';

    modalStack.push({ id: id, zIndex: zIndex, parentNode: domParentNode });
  },
  closeModal: function closeModal(id) {
    var _this = this;

    var modalStack = _popupContext2.default.getContext('modalStack');
    var modalDom = getModal();

    if (modalStack.length > 0) {
      var topItem = modalStack[modalStack.length - 1];
      if (topItem.id === id) {
        modalStack.pop();
        if (modalStack.length > 0) {
          modalDom.style.zIndex = modalStack[modalStack.length - 1].zIndex;
          modalDom.parentNode.removeChild(modalDom);
          var currModalParent = modalStack[0].parentNode;
          currModalParent && currModalParent.appendChild(modalDom);
        }
      } else {
        for (var i = modalStack.length - 1; i >= 0; i--) {
          if (modalStack[i].id === id) {
            modalStack.splice(i, 1);
            break;
          }
        }
      }
    }

    if (modalStack.length === 0) {
      setTimeout(function () {
        if (modalDom.parentNode) modalDom.parentNode.removeChild(modalDom);

        modalDom.style.display = 'none';
        _this.modalDom = null;
      }, 200);
    }
  }
};

exports.default = PopupManager;
