// zan-utils/url/readConfig.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports);
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports);
    global.readConfig = mod.exports;
  }
})(this, function (module, exports) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  var defaultConfig = {
    env: '',
    url: {
      imgqn: 'https://img.yzcdn.cn',
      cdn_static: 'https://b.yzcdn.cn/v2',
      env: 'online'
    }
  };

  /**
   * 读取 _global config
   * @memberof module:url
   * @param {string} key 需要从 _global 读取的字段
   * @return {any} 返回读取到的值
   * @example
   * readConfig('url.env');
   * // "online"
   */
  function readConfig(key) {
    var result = '';
    try {
      var readGlobalFunc = new Function('return window._global.' + key);
      result = readGlobalFunc();
    } catch (e) {}

    if (result) {
      return result;
    }

    try {
      var readDefaultFunc = new Function('var defaultConfig=' + JSON.stringify(defaultConfig) + ';return defaultConfig.' + key);
      result = readDefaultFunc();
    } catch (e) {}
    return result;
  }

  exports['default'] = readConfig;
  module.exports = exports['default'];
});
