// vant/lib/switch/index.js
'use strict';

exports.__esModule = true;

var _components;

var _loading = require('../loading');

var _loading2 = _interopRequireDefault(_loading);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  render: function render() {
    var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('div', { staticClass: "van-switch", class: ["van-switch--" + (_vm.checked ? 'on' : 'off'), { 'van-switch--disabled': _vm.disabled }], on: { "click": _vm.toggleState } }, [_c('div', { staticClass: "van-switch__node van-hairline-surround" }, [_vm.loading ? _c('van-loading', { staticClass: "van-switch__loading" }) : _vm._e()], 1), _c('div', { staticClass: "van-switch__bg" })]);
  },

  name: 'van-switch',
  components: (_components = {}, _components[_loading2.default.name] = _loading2.default, _components),
  props: {
    value: Boolean,
    disabled: Boolean,
    loading: Boolean,
    onChange: Function
  },
  data: function data() {
    return {
      checked: this.value
    };
  },

  watch: {
    checked: function checked(val) {
      this.$emit('input', val);
    },
    value: function value(val) {
      this.checked = val;
    }
  },
  methods: {
    /*
     * 开关状态交互。
     */
    toggleState: function toggleState() {
      if (this.disabled || this.loading) return;
      if (this.onChange) {
        this.onChange(!this.checked);
      } else {
        this.checked = !this.checked;
      }
    }
  }
};
