// zan-utils/validate/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './email', './number', './phone', './mobile', './postalCode', './date'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./email'), require('./number'), require('./phone'), require('./mobile'), require('./postalCode'), require('./date'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.email, global.number, global.phone, global.mobile, global.postalCode, global.date);
    global.index = mod.exports;
  }
})(this, function (module, exports, _email, _number, _phone, _mobile, _postalCode, _date) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _email2 = _interopRequireDefault(_email);

  var _number2 = _interopRequireDefault(_number);

  var _phone2 = _interopRequireDefault(_phone);

  var _mobile2 = _interopRequireDefault(_mobile);

  var _postalCode2 = _interopRequireDefault(_postalCode);

  var _date2 = _interopRequireDefault(_date);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = {
    email: _email2['default'],
    number: _number2['default'],
    phone: _phone2['default'],
    postalCode: _postalCode2['default'],
    date: _date2['default'],
    mobile: _mobile2['default']
  };
  module.exports = exports['default'];
});
