// vant/lib/swipe/utils.js
'use strict';

exports.__esModule = true;
exports.removeEvents = exports.bindEvents = exports.cancelAnimationFrame = exports.requestAnimationFrame = exports.EventEmitter = exports.extend = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _vue = require('vue');

var _vue2 = _interopRequireDefault(_vue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var extend = _assign2.default.bind(Object);

function EventEmitter() {
  this.__events = {};
}
EventEmitter.prototype = {
  on: function on(name, cb) {
    this.__events[name] || (this.__events[name] = []);
    this.__events[name].push(cb);
    return this;
  },
  emit: function emit(name) {
    var arr = this.__events[name];
    var argus = Array.prototype.slice.call(arguments, 1);
    var self = this;
    if (arr) {
      arr.forEach(function (cb) {
        cb.apply(self, argus);
      });
    }
  },
  removeListener: function removeListener(name, fn) {
    if (!this.__events[name]) {
      return;
    }
    var index = void 0;
    if (fn) {
      index = this.__events[name].indexOf(fn);
      if (index > 0) {
        this.__events[name].splice(index, 1);
      }
    } else {
      delete this.__events[name];
    }
  }
};

var isSupportRequestAnimationFrame = !_vue2.default.prototype.$isServer && (window.requestAnimationFrame || window.mozRequestAnimationFrame || window.webkitRequestAnimationFrame || window.msRequestAnimationFrame);
var isSupportCancelAnimationFrame = !_vue2.default.prototype.$isServer && (window.cancelAnimationFrame || window.mozCancelAnimationFrame || window.webkitCancelAnimationFrame || window.msCancelAnimationFrame);

var requestAnimationFrame = isSupportRequestAnimationFrame || function (callback, element) {
  return window.setTimeout(callback, 1000 / 60);
};

var cancelAnimationFrame = isSupportCancelAnimationFrame || function (id) {
  clearTimeout(id);
};

var bindEvents = function bindEvents(elem, eventNames, fn) {
  eventNames = eventNames.split(/\s+/);
  eventNames.forEach(function (eventName) {
    return elem.addEventListener(eventName, fn);
  });
};

var removeEvents = function removeEvents(elem, eventNames, fn) {
  eventNames = eventNames.split(/\s+/);
  eventNames.forEach(function (eventName) {
    return elem.removeEventListener(eventName, fn);
  });
};

exports.extend = extend;
exports.EventEmitter = EventEmitter;
exports.requestAnimationFrame = requestAnimationFrame;
exports.cancelAnimationFrame = cancelAnimationFrame;
exports.bindEvents = bindEvents;
exports.removeEvents = removeEvents;
