// zan-utils/browser/image.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './cookie'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./cookie'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.cookie);
    global.image = mod.exports;
  }
})(this, function (module, exports, _cookie) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _cookie2 = _interopRequireDefault(_cookie);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {
    return typeof obj;
  } : function (obj) {
    return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
  };

  function switchQiniuFormat(src, canWebp) {
    var pattern = /(\?imageView2\/\d\/w\/\d+\/h\/\d+\/q\/\d+\/format\/)(\w+)/;
    var result = void 0;

    result = src;

    if (pattern.test(src)) {
      var format = src.match(pattern)[2];
      if (canWebp) {
        if (format !== 'gif' && format !== 'webp') {
          result = src.replace(pattern, '$1webp');
        }
      } else if (format === 'webp') {
        // 对于不支持webp格式的浏览器把错误的地址切换回去
        result = src.replace(pattern, '$1jpg');
      }
    }

    return result;
  }
  /**
   * @memberof module:browser
   * @description
   * ##### 如果是新格式的图片则使用这种方式切换模式
   * 1、 image.toWebp(src)
   *  - 如果支持webp,给图片转化为新格式。类似`?imageView2/w/200/h/100/format/webp`
   *  - 如果不支持，转jpg
   *
   * 2、 image.checkCanWebp()
   *
   * 检查是否支持Webp格式
   */
  var image = {};
  image.toWebp = function () {
    var regex = /\.([^.!]+)!([0-9]{1,4})x([0-9]{1,4})q?([0-9]{0,2}|100)?(\+2x)?\..+/;
    var canWebP = false;

    try {
      canWebP = window.localStorage.getItem('canwebp') === 'ok';
    } catch (e) {}

    return function (_src) {
      var src = _src;
      var x2 = 1;

      var paras = src.match(regex);

      if (canWebP) {
        if (paras && paras.length >= 4) {
          // 类似于：!100x100.jpg这样的老格式
          // console.log(paras);
          // format/webp放最后面方便随时改格式
          if (paras[5] === '+2x') {
            x2 = 2;
          }
          src = src.replace(regex, '.') + paras[1] + '?imageView2/2/w/' + parseInt(paras[2], 10) * x2 + '/h/' + parseInt(paras[3], 10) * x2 + '/q/' + (paras[4] || 75) + '/format/' + (paras[1] === 'gif' ? 'gif' : 'webp');
        } else {
          // 如果不能匹配，则再判断一下图片地址是否为七牛格式的
          // 类似于：?imageView2/w/200/h/100/format/webp这样的新格式
          src = switchQiniuFormat(src, true);
        }
      } else if (paras && paras.length >= 4) {
        if (paras[5] === '+2x') {
          x2 = 2;
        }
        src = src.replace(regex, '.') + paras[1] + '?imageView2/2/w/' + parseInt(paras[2], 10) * x2 + '/h/' + parseInt(paras[3], 10) * x2 + '/q/' + (paras[4] || 75) + '/format/' + (paras[1] === 'webp' ? 'jpg' : paras[1]); // 如果不支持webp且是webp图片，则强制转换成jpg
      } else {
        src = switchQiniuFormat(src, false);
      }
      return src;
    };
  }();

  image.checkCanWebp = function () {
    // 检测是否支持webp
    var testWebP = function testWebP(callback) {
      var webP = new Image();
      webP.onload = webP.onerror = function () {
        // eslint-disable-line
        callback(webP.height === 2);
      };
      webP.src = 'data:image/webp;base64,UklGRjoAAAB' + 'XRUJQVlA4IC4AAACyAgCdASoCAAIALmk0mk0iIiIiIgBoSygABc6WWgAA/veff/0PP8bA//LwYAAA';
    };
    return function () {
      if (_typeof(window.localStorage) === 'object') {
        try {
          var canwebp = localStorage.getItem('canwebp');
          if (canwebp === 'ok') {
            (0, _cookie2['default'])('_canwebp', {
              value: '1',
              path: '/',
              domain: location.hostname,
              expires: 3650
            });
          } else if (canwebp !== 'no') {
            testWebP(function (canWebP) {
              localStorage.setItem('canwebp', canWebP ? 'ok' : 'no');
              if (canWebP) {
                (0, _cookie2['default'])('_canwebp', {
                  value: '1',
                  path: '/',
                  domain: location.hostname,
                  expires: 3650
                });
              }
            });
          }
        } catch (e) {}
      }
    };
  }();

  exports['default'] = image;
  module.exports = exports['default'];
});
