// zan-utils/browser/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './hasFlash', './cookie', './image', './isRetina', './local_storage', './session_storage', './ua_browser', './ua', './raf'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./hasFlash'), require('./cookie'), require('./image'), require('./isRetina'), require('./local_storage'), require('./session_storage'), require('./ua_browser'), require('./ua'), require('./raf'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.hasFlash, global.cookie, global.image, global.isRetina, global.local_storage, global.session_storage, global.ua_browser, global.ua, global.raf);
    global.index = mod.exports;
  }
})(this, function (module, exports, _hasFlash, _cookie, _image, _isRetina, _local_storage, _session_storage, _ua_browser, _ua, _raf) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _hasFlash2 = _interopRequireDefault(_hasFlash);

  var _cookie2 = _interopRequireDefault(_cookie);

  var _image2 = _interopRequireDefault(_image);

  var _isRetina2 = _interopRequireDefault(_isRetina);

  var _local_storage2 = _interopRequireDefault(_local_storage);

  var _session_storage2 = _interopRequireDefault(_session_storage);

  var _ua_browser2 = _interopRequireDefault(_ua_browser);

  var _ua2 = _interopRequireDefault(_ua);

  var _raf2 = _interopRequireDefault(_raf);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = {
    hasFlash: _hasFlash2['default'],
    isRetina: _isRetina2['default'],
    YZLocalStorage: _local_storage2['default'],
    sessionStorage: _session_storage2['default'],
    uaBrowser: _ua_browser2['default'],
    ua: _ua2['default'],
    cookie: _cookie2['default'],
    image: _image2['default'],
    raf: _raf2['default']
  };
  module.exports = exports['default'];
});
