// zan-utils/lib/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', './ajax', './fullfillImage', './local_storage', './hasFlash', './money/', './date/', './url/', './string/', './validate/', './browser', './number'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./ajax'), require('./fullfillImage'), require('./local_storage'), require('./hasFlash'), require('./money/'), require('./date/'), require('./url/'), require('./string/'), require('./validate/'), require('./browser'), require('./number'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.ajax, global.fullfillImage, global.local_storage, global.hasFlash, global.money, global.date, global.url, global.string, global.validate, global.browser, global.number);
    global.index = mod.exports;
  }
})(this, function (exports, _ajax, _fullfillImage, _local_storage, _hasFlash, _money, _date, _url, _string, _validate, _browser, _number) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.number = exports.browser = exports.validate = exports.string = exports.url = exports.date = exports.money = exports.hasFlash = exports.YZLocalStorage = exports.fullfillImage = exports.ajax = undefined;

  var _ajax2 = _interopRequireDefault(_ajax);

  var _fullfillImage2 = _interopRequireDefault(_fullfillImage);

  var _local_storage2 = _interopRequireDefault(_local_storage);

  var _hasFlash2 = _interopRequireDefault(_hasFlash);

  var _money2 = _interopRequireDefault(_money);

  var _date2 = _interopRequireDefault(_date);

  var _url2 = _interopRequireDefault(_url);

  var _string2 = _interopRequireDefault(_string);

  var _validate2 = _interopRequireDefault(_validate);

  var _browser2 = _interopRequireDefault(_browser);

  var _number2 = _interopRequireDefault(_number);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.ajax = _ajax2['default'];
  exports.fullfillImage = _fullfillImage2['default'];
  exports.YZLocalStorage = _local_storage2['default'];
  exports.hasFlash = _hasFlash2['default'];
  exports.money = _money2['default'];
  exports.date = _date2['default'];
  exports.url = _url2['default'];
  exports.string = _string2['default'];
  exports.validate = _validate2['default'];
  exports.browser = _browser2['default'];
  exports.number = _number2['default'];
});
