// zan-utils/string/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['module', 'exports', './addZero', './decodeHTMLEntities', './escape', './toCamelCase', './toSnakeCase', './unescape', './makeRandomString', './mapKeysToCamelCase', './mapKeysToSnakeCase'], factory);
  } else if (typeof exports !== "undefined") {
    factory(module, exports, require('./addZero'), require('./decodeHTMLEntities'), require('./escape'), require('./toCamelCase'), require('./toSnakeCase'), require('./unescape'), require('./makeRandomString'), require('./mapKeysToCamelCase'), require('./mapKeysToSnakeCase'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod, mod.exports, global.addZero, global.decodeHTMLEntities, global.escape, global.toCamelCase, global.toSnakeCase, global.unescape, global.makeRandomString, global.mapKeysToCamelCase, global.mapKeysToSnakeCase);
    global.index = mod.exports;
  }
})(this, function (module, exports, _addZero, _decodeHTMLEntities, _escape, _toCamelCase, _toSnakeCase, _unescape, _makeRandomString, _mapKeysToCamelCase, _mapKeysToSnakeCase) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _addZero2 = _interopRequireDefault(_addZero);

  var _decodeHTMLEntities2 = _interopRequireDefault(_decodeHTMLEntities);

  var _escape2 = _interopRequireDefault(_escape);

  var _toCamelCase2 = _interopRequireDefault(_toCamelCase);

  var _toSnakeCase2 = _interopRequireDefault(_toSnakeCase);

  var _unescape2 = _interopRequireDefault(_unescape);

  var _makeRandomString2 = _interopRequireDefault(_makeRandomString);

  var _mapKeysToCamelCase2 = _interopRequireDefault(_mapKeysToCamelCase);

  var _mapKeysToSnakeCase2 = _interopRequireDefault(_mapKeysToSnakeCase);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = {
    addZero: _addZero2['default'],
    decodeHTMLEntities: _decodeHTMLEntities2['default'],
    escape: _escape2['default'],
    toCamelCase: _toCamelCase2['default'],
    toSnakeCase: _toSnakeCase2['default'],
    unescape: _unescape2['default'],
    makeRandomString: _makeRandomString2['default'],
    mapKeysToCamelCase: _mapKeysToCamelCase2['default'],
    mapKeysToSnakeCase: _mapKeysToSnakeCase2['default']
  };
  module.exports = exports['default'];
});
